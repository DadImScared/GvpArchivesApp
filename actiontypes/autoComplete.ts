export const ADD_AUTOCOMPLETE = "autoComplete/ADD_AUTOCOMPLETE";
