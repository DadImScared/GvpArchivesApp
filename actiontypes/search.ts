export const UPDATE_SEARCH_QUERY = "search/UPDATE_SEARCH_QUERY";
