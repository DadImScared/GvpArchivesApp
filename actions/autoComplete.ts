import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/autoComplete";

const addAutoComplete = (query: string, categories: string[], results: string[]) => {
  return action(ActionTypes.ADD_AUTOCOMPLETE, { query, categories, results });
};

export default {
  addAutoComplete
};
