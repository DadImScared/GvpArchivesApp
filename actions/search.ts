import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
