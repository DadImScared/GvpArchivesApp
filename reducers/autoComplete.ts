import moment from "moment";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import autoComplete from "../actions/autoComplete";
import * as ActionTypes from "../actiontypes/autoComplete";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { ILoadingState } from "./loading";
import { getQueryId, getSearch, ISearchState } from "./search";

export interface IAutoCompleteState {
  [key: string]: {
    results: string[];
    timeToLive: number;
  };
}

export const getInitialAutoCompleteState = () => ({});

export default function(state = getInitialAutoCompleteState(), action: ActionType<typeof autoComplete>) {
  switch (action.type) {
    case ActionTypes.ADD_AUTOCOMPLETE:
      const { payload: { query, categories, results } } = action;
      return {
        ...state,
        [getQueryId(query, categories)]: {
          results,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    default:
      return state;
  }
}

interface IGetAutoCompleteReturn {
  results: string[];
  timeToLive: number;
}

export interface IAutoCompleteScreenData {
  autoComplete: IGetAutoCompleteReturn;
  query: string;
  queryId: string;
  categories: string[];
  isLoading: boolean;
}

export const getAutoCompleteScreen = createSelector<
  IReducerState,
  IAutoCompleteState,
  ISearchState,
  ILoadingState,
  IAutoCompleteScreenData
>(
  [(state) => state.autoComplete, getSearch, (state) => state.loading],
  (autoCompleteState, { categories, query }, loadingState) => {
    const queryId = getQueryId(query, categories);
    return {
      categories,
      query,
      queryId,
      autoComplete: autoCompleteState[queryId] || { results: [], timeToLive: 0 },
      isLoading: loadingState[`${AUTOCOMPLETE_QUERY}${queryId}`]
    };
  }
);
