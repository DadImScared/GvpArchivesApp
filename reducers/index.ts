import _ from "lodash";
import { combineReducers, DeepPartial } from "redux";

import audioPlayer, { getInitialAudioPlayerState, IAudioPlayerState } from "./audioPlayer";
import autoComplete, { getInitialAutoCompleteState, IAutoCompleteState } from "./autoComplete";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";
import search, { getInitialSearchState, ISearchState } from "./search";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  autoComplete: IAutoCompleteState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
  search: ISearchState;
}

export const getInitialReducerState = (injectedState: DeepPartial<IReducerState> = {}) => _.merge({
  audioPlayer: getInitialAudioPlayerState(),
  autoComplete: getInitialAutoCompleteState(),
  itemsByCategory: {},
  itemsById: {},
  loading: {},
  search: getInitialSearchState()
}, injectedState);

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  autoComplete,
  itemsByCategory,
  itemsById,
  loading,
  search
});

export default reducer;
