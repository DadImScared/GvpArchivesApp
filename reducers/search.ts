import _ from "lodash";
import { ActionType } from "typesafe-actions";

import { search } from "../actions";
import * as ActionTypes from "../actiontypes/search";
import { IReducerState } from "./index";

export interface ISearchState {
  categories: string[];
  query: string;
}

export type SearchStateAndActions<
  T extends keyof ISearchState | keyof typeof search
  > = Pick<ISearchState & typeof search, T>;

export const CATEGORIES = [
  "bhagavatpatrika",
  "book",
  "harikatha",
  "harmonistmonthly",
  "harmonistmagazine",
  "lecture",
  "movie",
  "song"
];

export const getInitialSearchState = (): ISearchState => ({
  categories: CATEGORIES,
  query: ""
});

export default function(state = getInitialSearchState(), action: ActionType<typeof search>) {
  switch (action.type) {
    case ActionTypes.UPDATE_SEARCH_QUERY:
      return {
        ...state,
        query: action.payload
      };
    default:
      return state;
  }
}

export const getQueryId = (query: string, categories: string[]) => {
  return `${query}-${_.sortBy(categories, (o) => o).join(",")}`;
};

export const getSearch = (state: IReducerState) => state.search;
