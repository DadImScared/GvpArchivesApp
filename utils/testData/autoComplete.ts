import _ from "lodash";

import { getQueryId } from "../../reducers/search";
import { getMockFunction } from "./search";

export const autoCompleteTestData = ({ props: injectedProps = {}, storybook = false } = {}) => {
  const props = {
    autoComplete: {
      results: ["suggestion-1", "suggestion-2", "suggestion-3", "suggestion-4", "suggestion-5"],
      timeToLive: 1234234
    },
    categories: ["book"],
    isLoading: false,
    navigation: {
      dangerouslyGetParent: () => props.navigation,
      replace: getMockFunction({ storybook, actionName: "replace" })
    } as any,
    query: "radha",
    queryId: getQueryId("radha", ["book"]),
    updateSearchQuery: getMockFunction({ storybook, actionName: "updateSearchQuery" })
  };
  return _.merge(props, injectedProps);
};
