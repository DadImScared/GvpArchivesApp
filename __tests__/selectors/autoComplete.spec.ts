import { getInitialReducerState } from "../../reducers";
import { getAutoCompleteScreen } from "../../reducers/autoComplete";
import { CATEGORIES, getQueryId } from "../../reducers/search";

describe("autoComplete", () => {
  it("should get auto complete screen", () => {
    const state = getInitialReducerState({
      autoComplete: {
        [getQueryId("my query", CATEGORIES)]: {
          results: ["text"],
          timeToLive: 1555834576011
        }
      },
      search: {
        query: "my query"
      }
    });
    expect(getAutoCompleteScreen(state)).toMatchSnapshot();
  });
});
