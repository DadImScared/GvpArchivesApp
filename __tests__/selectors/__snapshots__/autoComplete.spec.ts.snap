// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoComplete should get auto complete screen 1`] = `
Object {
  "autoComplete": Object {
    "results": Array [
      "text",
    ],
    "timeToLive": 1555834576011,
  },
  "categories": Array [
    "bhagavatpatrika",
    "book",
    "harikatha",
    "harmonistmonthly",
    "harmonistmagazine",
    "lecture",
    "movie",
    "song",
  ],
  "isLoading": undefined,
  "query": "my query",
  "queryId": "my query-bhagavatpatrika,book,harikatha,harmonistmagazine,harmonistmonthly,lecture,movie,song",
}
`;
